File: app/src/test/java/com/pomo/sync/recovery/ReplicaRecoveryTest.kt
package com.pomo.sync.recovery

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

public class ReplicaRecoveryTest {
    @Test
    public fun unionsVerifiedSourcesWithProvenanceAndListsExactGaps() {
        val checkpoint = CheckpointManifest(
            "cp",
            "RECOVERY",
            listOf(FrontierHead("a", 1, "a1")),
            1,
            "root",
            listOf("pack"),
            emptyList(),
        )
        val a1 = PackedOperation("a1", "a", 1, byteArrayOf(1))
        val a3 = PackedOperation("a3", "a", 3, byteArrayOf(3))
        val plan = Rehydrator.plan(
            listOf(
                RecoverySource("device", checkpoint, listOf(a1)),
                RecoverySource("mailbox", checkpoint, listOf(a1, a3)),
            ),
        )
        assertEquals(setOf("device", "mailbox"), plan.sourceByOperation.getValue("a1"))
        assertEquals(setOf("a@2"), plan.gaps)
        assertEquals(listOf("a1", "a3"), plan.operations.map { it.operationId })
    }

    @Test
    public fun packsOnlyReplaceCompleteForkFreePrefixesAndNamedAnchorsRemainInspectable() {
        val pack = JournalPack(
            "pack",
            FrontierHead("a", 2, "a2"),
            listOf(
                PackedOperation("a1", "a", 1, byteArrayOf(1)),
                PackedOperation("a2", "a", 2, byteArrayOf(2)),
            ),
        )
        CheckpointPolicy.validatePack(pack, emptySet())
        assertTrue(runCatching { CheckpointPolicy.validatePack(pack, setOf("a")) }.isFailure)
        assertTrue(
            runCatching {
                CheckpointPolicy.validate(CheckpointManifest("cp", "SAFETY", emptyList(), 1, "root", emptyList(), emptyList()))
            }.isSuccess,
        )
    }

    @Test
    public fun projectionCorruptionRebuildsButJournalOrKeyCorruptionSealsReplica() {
        assertFalse(integrityDisposition(IntegrityFailure.PROJECTION_CORRUPT).active)
        assertTrue(integrityDisposition(IntegrityFailure.JOURNAL_CORRUPT).incarnationSealed)
        assertTrue(integrityDisposition(IntegrityFailure.DEVICE_KEY_MISSING).inspectionAllowed)
    }
}

File: extension/tests/sync/replicaRecovery.test.ts
import { describe, expect, test } from "bun:test";
import { integrityDisposition, planRehydration, validateCheckpoint, validateJournalPack } from "../../src/sync/recovery/replicaRecovery";

describe("Checkpoints, packs, rehydration, and Safe mode", () => {
  const checkpoint = { checkpointId: "cp", kind: "RECOVERY" as const, frontier: [{ feedKey: "a", sequence: 1, operationId: "a1" }], materializerVersion: 1, projectionRoot: "root", packIds: ["pack"], blobIds: [] };
  const a1 = { operationId: "a1", feedKey: "a", sequence: 1, wire: new Uint8Array([1]) };
  const a3 = { operationId: "a3", feedKey: "a", sequence: 3, wire: new Uint8Array([3]) };

  test("unions sources without inventing missing Operations and keeps provenance", () => {
    const plan = planRehydration([
      { sourceId: "device", checkpoint, operations: [a1] },
      { sourceId: "mailbox", checkpoint, operations: [a1, a3] },
    ]);
    expect(plan.sourceByOperation.get("a1")).toEqual(new Set(["device", "mailbox"]));
    expect(plan.gaps).toEqual(new Set(["a@2"]));
  });

  test("accepts named anchors and only packs complete fork-free prefixes", () => {
    validateCheckpoint({ ...checkpoint, kind: "SAFETY" });
    const pack = { packId: "pack", prefix: { feedKey: "a", sequence: 2, operationId: "a2" }, operations: [a1, { operationId: "a2", feedKey: "a", sequence: 2, wire: new Uint8Array([2]) }] };
    validateJournalPack(pack, new Set());
    expect(() => validateJournalPack(pack, new Set(["a"]))).toThrow(/replaceable/);
  });

  test("rejects incomparable checkpoint frontiers instead of selecting one", () => {
    const other = { ...checkpoint, checkpointId: "other", frontier: [{ feedKey: "b", sequence: 2, operationId: "b2" }] };
    expect(() => planRehydration([
      { sourceId: "first", checkpoint, operations: [] },
      { sourceId: "second", checkpoint: other, operations: [] },
    ])).toThrow(/incomparable/);
  });

  test("rebuilds projections but seals the incarnation for journal or key corruption", () => {
    expect(integrityDisposition("PROJECTION_CORRUPT")).toEqual({ active: false, incarnationSealed: false, inspectionAllowed: true, exportAllowed: true });
    expect(integrityDisposition("JOURNAL_CORRUPT").incarnationSealed).toBeTrue();
    expect(integrityDisposition("DEVICE_KEY_MISSING").inspectionAllowed).toBeTrue();
  });
});
